diff --git a/src/Data/Serialize/Get.hs b/src/Data/Serialize/Get.hs
index 87fd159..c08ee3e 100644
--- a/src/Data/Serialize/Get.hs
+++ b/src/Data/Serialize/Get.hs
@@ -741,7 +741,11 @@ foreign import ccall unsafe "stg_uncheckedShiftL64"
 #endif
 
 #else
+# if __GLASGOW_HASKELL__ >= 903
+shiftl_w64 (W64# w) (I# i) = W64# (w `uncheckedShiftL64#` i)
+# else
 shiftl_w64 (W64# w) (I# i) = W64# (w `uncheckedShiftL#` i)
+# endif
 #endif
 
 #else
