diff --git a/Control/Concurrent/Async/Pool/Async.hs b/Control/Concurrent/Async/Pool/Async.hs
index be4e7c0..7d189d1 100644
--- a/Control/Concurrent/Async/Pool/Async.hs
+++ b/Control/Concurrent/Async/Pool/Async.hs
@@ -711,9 +711,21 @@ tryAll = try
 {-# INLINE rawForkIO #-}
 rawForkIO :: IO () -> IO ThreadId
 rawForkIO action = IO $ \ s ->
-   case (fork# action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
+   case (fork#
+#if __GLASGOW_HASKELL__ >= 903
+           (unIO action)
+#else
+           action
+#endif
+           s) of (# s1, tid #) -> (# s1, ThreadId tid #)
 
 {-# INLINE rawForkOn #-}
 rawForkOn :: Int -> IO () -> IO ThreadId
 rawForkOn (I# cpu) action = IO $ \ s ->
-   case (forkOn# cpu action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
+   case (forkOn# cpu
+#if __GLASGOW_HASKELL__ >= 903
+           (unIO action)
+#else
+           action
+#endif
+           s) of (# s1, tid #) -> (# s1, ThreadId tid #)
