diff --git a/relude.cabal b/relude.cabal
index c8d6233..a7f9599 100644
--- a/relude.cabal
+++ b/relude.cabal
@@ -225,7 +225,7 @@ library
                      , containers >= 0.5.7 && < 0.7
                      , deepseq ^>= 1.4
                      , ghc-prim >= 0.4.0.0 && < 0.8
-                     , hashable >= 1.2 && < 1.4
+                     , hashable >= 1.2 && < 1.5
                      , mtl ^>= 2.2
                      , stm >= 2.4 && < 2.6
                      , text ^>= 1.2
diff --git a/src/Relude/Monoid.hs b/src/Relude/Monoid.hs
index 0456c4a..411ba9f 100644
--- a/src/Relude/Monoid.hs
+++ b/src/Relude/Monoid.hs
@@ -37,8 +37,12 @@ import Data.Monoid (Ap (..))
 #endif
 import Data.Monoid (All (..), Alt (..), Any (..), Dual (..), Endo (..), First (..), Last (..),
                     Monoid (..), Product (..), Sum (..))
-import Data.Semigroup (Option (..), Semigroup (sconcat, stimes, (<>)), WrappedMonoid, cycle1,
-                       mtimesDefault, stimesIdempotent, stimesIdempotentMonoid, stimesMonoid)
+import Data.Semigroup (Semigroup (sconcat, stimes, (<>)), WrappedMonoid, cycle1,
+                       mtimesDefault, stimesIdempotent, stimesIdempotentMonoid, stimesMonoid
+#if !MIN_VERSION_base(4,16,0)
+                       , Option(..)
+#endif
+                      )
 
 import Relude.Bool.Reexport (Bool (..))
 import Relude.Monad.Reexport (Maybe, fromMaybe)
